src/variable.rs

```diff
@@ -66,580 +217,423 @@ pub trait Numeric {
-    /// All slice elements must be > 0.
-    pub fn values_at<T: Numeric>(&self, indices: &[usize], slice_len: &[usize]) -> Result<Vec<T>, String> {
-        T::slice_from_variable(self, indices, slice_len)
-    }
+    #[cfg(feature = "ndarray")]
+    /// Fetches variable
+    pub fn values<T: Numeric>(
+        &self,
+        indices: Option<&[usize]>,
+        size_len: Option<&[usize]>,
+    ) -> error::Result<ArrayD<T>> {
+        let _indices: Vec<usize>;
+        let indices = match indices {
+            Some(x) => {
+                self.check_indices(x, false)?;
+                x
+            }
+            None => {
+                _indices = self.default_indices(false)?;
+                &_indices
+            }
+        };
+        let _size_len: Vec<usize>;
+        let full_length = self.dimensions.iter().map(|d| d.len()).product::<usize>();
+        let size_len = match size_len {
+            Some(x) => {
+                self.check_sizelen(full_length, indices, x, false)?;
+                x
+            }
+            None => {
+                _size_len = self.default_sizelen(full_length, indices, false)?;
+                &_size_len
+            }
+        };
 
-    /// Fetchs variable values as a ndarray.
-    ///
-    /// ```
-    /// // Each values will be implicitly casted to a f64 if needed
-    /// // let values: ArrayD<f64> = some_variable.as_array().unwrap();
-    /// ```
-    ///
-    pub fn as_array<T: Numeric>(&self) -> Result<ArrayD<T>, Box<Error>> {
-        let mut dims: Vec<usize> = Vec::new();
-        for dim in &self.dimensions {
-            dims.push(dim.len as usize);
+        let mut values = Vec::with_capacity(full_length);
+        unsafe {
+            T::variable_to_ptr(self, indices, size_len, values.as_mut_ptr())?;
+            values.set_len(full_length);
         }
-        let values = self.values()?;
-        Ok(ArrayD::<T>::from_shape_vec(dims, values)?)
+        Ok(ArrayD::from_shape_vec(size_len, values).unwrap())
     }
-    
-    /// Fetchs variable slice as a ndarray.
-    pub fn array_at<T: Numeric>(&self, indices: &[usize], slice_len: &[usize]) -> Result<ArrayD<T>, Box<Error>> {
-        let values = self.values_at(indices, slice_len)?;
-        Ok(ArrayD::<T>::from_shape_vec(slice_len, values)?)
+
+    /// Fetches variable into slice
+    /// buffer must be able to hold all the requested elements
+    pub fn values_to<T: Numeric>(
+        &self,
+        buffer: &mut [T],
+        indices: Option<&[usize]>,
+        size_len: Option<&[usize]>,
+    ) -> error::Result<()> {
+        let _indices: Vec<usize>;
+        let indices = match indices {
+            Some(x) => {
+                self.check_indices(x, false)?;
+                x
+            }
+            None => {
+                _indices = self.default_indices(false)?;
+                &_indices
+            }
+        };
+        let _size_len: Vec<usize>;
+        let size_len = match size_len {
```

Comment:
confusing variable name, shouldn't this be `slice_len` since the corresponding parameter to variable_to_ptr is named like that? :) Or slice boundaries.

Reply:
Changed to `slice_len` for consistency